File: windows/PrestreloAjuda/Services/TeamsConfig.cs
using System.IO;
using System.Text.Json;

namespace PrestreloAjuda.Services;

/// Um time jogável (define qual conjunto de soluções a Elite 4 usa).
public sealed class TeamInfo
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    /// Sprite usado como ícone do time (data/sprites/<icon>.png).
    public string? Icon { get; set; }
    public List<string> Pokemon { get; set; } = new();
    public string? Pokepaste { get; set; }
    /// CODE do Pokeking pra extrair o time no jogo (mostrado no botão de fonte).
    public string? Code { get; set; }
    /// Se o time tem versão "Modo Emoji" (combate na notação original do autor).
    public bool HasEmoji { get; set; }
}

/// Manifesto data/teams.json: lista de times + quais modos variam por time.
public sealed class TeamsConfig
{
    public string Default { get; set; } = "";
    public List<string> TeamScopedModes { get; set; } = new();
    public List<TeamInfo> Teams { get; set; } = new();

    private static readonly JsonSerializerOptions Opts = new() { PropertyNameCaseInsensitive = true };

    public TeamInfo? Get(string id) => Teams.FirstOrDefault(t => t.Id == id);
    public bool IsTeamScoped(string modeId) => TeamScopedModes.Contains(modeId);

    public TeamInfo Resolve(string? id) =>
        (id != null ? Get(id) : null) ?? Get(Default) ?? Teams.FirstOrDefault()
        ?? new TeamInfo { Id = "default", Name = "Padrão" };

    public static TeamsConfig Load()
    {
        var path = Path.Combine(SolveLoader.DataDir, "teams.json");
        try
        {
            if (File.Exists(path))
                return JsonSerializer.Deserialize<TeamsConfig>(File.ReadAllText(path), Opts) ?? Fallback();
        }
        catch { /* cai no fallback */ }
        return Fallback();
    }

    /// Sem manifesto: um único "time" que lê os elite4 da raiz (compat retro).
    private static TeamsConfig Fallback() => new()
    {
        Default = "",
        TeamScopedModes = new(),
        Teams = new() { new TeamInfo { Id = "", Name = "Padrão" } }
    };
}

File: android/app/src/main/java/com/reload/prestreloajuda/data/TeamsConfig.kt
package com.reload.prestreloajuda.data

import android.content.Context
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** Um time jogável (define qual conjunto de soluções a Elite 4 usa). */
@Serializable
data class TeamInfo(
    val id: String = "",
    val name: String = "",
    /** Sprite usado como ícone do time (assets/data/sprites/<icon>.png). */
    val icon: String? = null,
    val pokemon: List<String> = emptyList(),
    val pokepaste: String? = null,
    /** CODE do Pokeking pra extrair o time no jogo (mostrado no botão de fonte). */
    val code: String? = null,
    /** Se o time tem versão "Modo Emoji" (combate na notação original do autor). */
    val hasEmoji: Boolean = false,
)

/** Manifesto data/teams.json: lista de times + quais modos variam por time. */
@Serializable
data class TeamsConfig(
    @SerialName("default") val defaultTeam: String = "",
    val teamScopedModes: List<String> = emptyList(),
    val teams: List<TeamInfo> = emptyList(),
) {
    fun get(id: String?): TeamInfo? = id?.let { i -> teams.firstOrNull { it.id == i } }
    fun isTeamScoped(modeId: String) = teamScopedModes.contains(modeId)
    fun resolve(id: String?): TeamInfo =
        get(id) ?: get(defaultTeam) ?: teams.firstOrNull() ?: TeamInfo(id = "", name = "Padrão")

    companion object {
        private val json = Json { ignoreUnknownKeys = true; isLenient = true }

        fun load(context: Context): TeamsConfig = try {
            val text = context.assets.open("data/teams.json")
                .bufferedReader(Charsets.UTF_8).use { it.readText() }
            json.decodeFromString(serializer(), text)
        } catch (e: Exception) {
            // Sem manifesto: um único "time" que lê os elite4 da raiz (compat retro).
            TeamsConfig(teams = listOf(TeamInfo(id = "", name = "Padrão")))
        }
    }
}
